### ISSUE TITLE
[java] Replace UseConcurrentHashMap with UseConcurrentCollection

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#2926]
**Is your feature request related to a problem? Please describe.**

The rule [UseConcurrentHashMap](https://pmd.github.io/latest/pmd_rules_java_multithreading.html#useconcurrenthashmap) is currently very noisy, as it will flag any local variable or field declaration that looks like so:
```
Map<...> map = new HashMap<>();
```

Many collections don't need synchronized access, which is why ArrayList is so much more popular than Vector. So this rule goes over the top by requiring that all maps be concurrent (and failing to do so, because it only considers variable initializers).

**Describe the solution you'd like**

I think the rule should actually flag places where there would be a clear benefit from using a concurrent map. Eg improper usages of `Collections.synchronizedMap`, like the following

```java
// warn, ConcurrentHashMap would be more adapted
Collections.synchronizedMap(new HashMap<>())

 // do not warn, this is actually the correct use case for synchronizedMap
Collections.synchronizedMap(aMapThatComesFromSomewhereElse)
```

Obviously it should not be restricted to looking in variable initializers.

Another possible culprit would be using `Collections.synchronizedSet(new HashSet<>())`. There is no "ConcurrentSet" interface, but ConcurrentMap is more efficient than the synchronized set, eg because it locks on individual buckets. This is why we sometimes use a ConcurrentMap instead of a Set in our codebase, eg [here](https://github.com/pmd/pmd/blob/bc4938087a9538c96134ecdff560f987aac22981/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/PMDASMClassLoader.java#L45-L48). Also [refs this discussion](https://github.com/pmd/pmd/pull/1006#discussion_r179967789)

I found out recently that there is a [`Collections.newSetFromMap`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#newSetFromMap-java.util.Map-) method. This allows creating a `Set` using whatever implementation of Map you want, including concurrent maps.

To sum it up, I think we should replace the rule with a new `UseConcurrentCollection` rule, which looks for
  * `synchronizedMap(new HashMap<>())` -> replace with `new ConcurrentHashMap<>()`
  * `synchronizedSortedMap(new TreeMap<>())` -> replace with `new ConcurrentSkipListMap<>()`
  * `synchronizedSet(new HashSet<>())` -> replace with `Collections.newSetFromMap(new ConcurrentHashMap<>())`
  * `synchronizedSet(new TreeSet<>())` -> replace with `Collections.newSetFromMap(new ConcurrentSkipListMap<>())`
  * similarly with `synchronized[Sorted|Navigable][Set|Map]`

**Describe alternatives you've considered**
* Make separate rules for sets and maps. I think this would not be useful, because they're so closely related. Also maybe in the future we can extend it to new patterns too, thanks to the generic name

**Additional context**

Current issues with `UseConcurrentHashMap` are all complaining of the noise: #2786, #2201, #1022

Maybe in the future we can have another rule to try to detect maps that are used as sets, and suggest replacing them with `newSetFromMap`

**Comment by numeralnathan** [comment]
I agree that this rule is too noisy and gives bad recommendations.  I have this rule disabled in all of my projects.

However, replacing `synchronized...` with `new Concurrent...` is not always a good idea either.  `synchronizedMap(new HashMap<>())` consumes much less heap space than `new ConcurrentHashMap<>()` yet could cause a bottleneck.  It really depends on the use case.  If heap usage is the constraining factor in a server, then using a synchronized `Map` would be the better choice.  On the other hand, if a synchronized `Map` is a bottleneck, then a `ConcurrentHashMap`.

Since it is difficult to know which way to recommend, I suggest removing this rule altogether unless someone can figure out how to determine the best course of action from a static analysis.

**Comment by oowekyala** [comment]
>  It really depends on the use case. If heap usage is the constraining factor in a server, then using a synchronized Map would be the better choice.

I think the problems you underline are valid, but this tradeoff should anyway be considered by a user when including the rule, as part of the "is this rule right for my codebase" question. If their understanding is that they would *never* use a concurrent map over a synchronized map, then they should exclude the rule. If it's more blurry, and they think the rule may be useful, but have decided that in some instances it's better to keep a synchronized map, then suppressing those specific violations with a comment explaining why could provide some useful documentation to a reader of the code.

So the tradeoff between memory usage & concurrency could be explained in the documentation for the user to make an informed decision. But I don't think the mere fact that there is a tradeoff means the rule is inappropriate in all cases.

If I'm wrong, and this new rule is useless, then I agree that removing the rule would anyway be better than keeping it in

**Comment by numeralnathan** [comment]
Perhaps, we should add a parameter to the rule (e.g. `heapConstrained`).  The user knows if their program is heap constrained or scalability constrained.  The rule can then generally provide the right recommendation for the user's program.

If the user sets the parameter to `true`, then the rule recommends using synchronized data structures.  If `false`, then the rule recommends using concurrent data structures.  Of course, there will be exceptions to this general recommendation and the user will need to select the opposite kind of multi-threaded access.  The user can then add `// NOPMD - Scalability bottleneck` or `// NOPMD - Memory hog` as the case may be.

What do you think?

**Comment by Bukama** [comment]
I would like to upvote this after I had a false positive yesterday.


This is fine:
```
Map<String, String> myMap = new ConcurrentHashMap<>();
```

This is a violation, even that it's also a thread safe concurrent Map

```
Map<String, String> myMap = new ConcurrentSkipListMap<>();
```

This again is fine
```
ConcurrentMap<String, String> myMap = new ConcurrentSkipListMap<>();
```
